rasa/cli/initial_project/config.yml

```diff
@@ -1,26 +1,9 @@
 # Configuration for Rasa NLU.
 # https://rasa.com/docs/rasa/nlu/components/
 language: en
+
 pipeline:
```

Comment:
I think to avoid confusion, the keys `pipeline` and `policies` should be removed completely

Reply:
I am actually not sure what is more / less confusing. I find it quite intuitive to say: These (`language`, `pipeline`, `policies`) are the keys a configuration file needs. If you want one of them to be configured automatically, just leave it empty.

Implementation-wise, having it like this is a lot easier, because otherwise it gets tricky where exactly to put the new sections with respect to the already existing sections and with respect to comments that might or might not exist in the `config.yml`.

I have discussed this with @wochinge already, we thought it might be good to share with the CSE and engineering teams and gather feedback on how the feature would be easy and intuitive to use. What do you think?

Reply:
I think from a users perspective it is better to not have these keys. Here is the most common  interaction users will go through:
- they will run `rasa init`. for a short period, the config will only contain `languge: en`
- most likely they will train a model right after, probably even as part of `rasa init`. as soon as that is done, the config will contain the auto-configuration.
- if they want to customize this config, they only need to remove the comment

If you keep the keys, users will end up with something like this:
```yml
language: en

# some comment ...
pipeline:

# more comments ...
policies:

# ****our marker
# ... all the instructions we came up with ...
# policies:
# - ...
# - ...
# pipeline:
# - ...
# - ...
```
which means they are most likely going to run into the issue where they uncoment the autoconfiguration, but miss to remove the above dangling keys resulting in duplicated yaml dict keys. I think this is valid yaml, meaning it can result in very odd behaviour from a users perspective e.g. if they start to modify the first `pipeline` key but don't see any change since the second key is used.

Reply:
The way I've implemented it currently, the keys are always present exactly once. `rasa init` also adds those empty keys, and the auto configuration always checks if the key is empty. If it isn't, no auto configuration is run. If it is, auto configuration is run and the empty section is filled with stuff. This is independent of any added marker, only based on presence of content in the section.

I've also made sure now that the auto configured content is commented out.
A problem that I still need to fix: When running auto configuration multiple times, the auto configured content will now be duplicated multiple times (old comments get an additional `#` sign and are kept, new ones are added).

Reply:
So basically, what users would end up with (once the problem I mentioned is fixed) is something like this:

```
language: en

# some comment ...
# **** our marker
# ... all the instructions we came up with followed by the autoconfiguration output for pipeline...
pipeline:
# - ...
# - ...

# more comments ...
# **** our marker
# ... all the instructions we came up with followed by the autoconfiguration output for policies...
policies:
# - ...
# - ...
```

Reply:
mhm this seems harder to implement without a user added benefit: e.g. the issue you are running into with the duplication is because the markers are useless as you can't use them to replace the whole auto config block - I am not sure you need them at all in your version. 

your decision though, I think both versions are fine.

Reply:
Hm... I still see issues with both versions:

My version:
- the duplication -> My plan to fix this is: for a key that is empty (so where auto config will run), remove all comments immediately below it, i.e. until the next empty line, before dumping the new automatic content
- That means that actual comments put there by the user will get deleted, too. So basically, you can't put comments right into a section that will be auto configured.

Your version:
- It works fine if both keys are commented in or both commented out. But what if the user lets auto config run once, then manually changes only `policies`, and then trains again? Then we mustn't overwrite the whole section. So maybe we'd need two markers, one for `policies`, one for `pipeline`?

Reply:
The feedback (yours and Ben's, also Tobi's initial reaction) has convinced me that it would be useful if `pipeline` and `policies` can also be left out completely.

The objection I had against that was that it's tricky / impossible to choose where to position the keys with respect to existing comments: I want to allow users to do something like 
```
# Configuration for Rasa NLU
language: en

# some notes on pipeline
# ...

# Configuration for Rasa Core
```
and have the keys for `pipeline` and `policies` show up where they want them.

So my proposed solution now is: Allow users both to leave a key out completely or to add a key and leave it empty. If present but empty, the auto config will leave the section where it is and just fill it with (commented out) content. If a key is not present at all, it will also be auto configured, and the section will be added to the end of the config file below everything else.

I think in the current setup (without markers) that should work easily and not cause confusion.
@tmbo What do you think? Do you see any issues with that approach?
(cc @wochinge)

Reply:
I do not want to support multiple ways, there is no benefit in that for this feature. 

I don't get the issues around positioning tbh. Configuration for missing keys should just be added as a comment to the end. if the file already contains a comment with the magic markers, that comment will be removed first. If you want to get fancy, you can replace in place, but even that is not necessary.

Don't make this feature more complicated than it needs to be, the logic should just be:
1. check if `policy` is in the config -> if not, remember that
2. check if `pipeline` is in the config -> if not, remember that
3. check the configuration for the auto config section and remove it
4. append a new autoconfiguration comment to the end of the file with all the keys that where missing

Reply:
That's how I summarized it for myself. Imo it doesn't matter which approach we take:

- **Approach 1:** Dump auto config as one block at the end
    - what if the keys are there?

        → we'd have to remove the empty ones before dumping the comments

    - what if only 1 of 2 keys needs to be auto configured? → doesn't matter, we delete the whole block and add the missing one commented out at the end.
    - **We have to make sure that comments are commented out twice, so when commenting out the whole block, the actual comments stay comments**
- **Approach 2:** Dump auto configuration below each key (once for `policies` , once for `pipeline`)
    - what if the keys are not there? → add the keys + commented out configs
    - what if only 1 of 2 keys needs to be auto configured? → doesn't matter
- Other thoughts
    - Which case is more likely? → Probably whatever `rasa init` suggests. Imo we should be able to deal with both.
    - What if new keys are added? → not a problem with both approaches

Reply:
Agree :+1:

as I said earlier in this thread
> your decision though, I think both versions are fine.

Strong no on supporting both at the same time.

Reply:
Supporting both would be a minor change of the current version: When a key is not present, it is added as an empty key at the end of the file before auto configuration is run. Why do you see this as problematic? (Let's maybe talk about this in a 10 min call? that's probably quicker that typing out a bunch of comments)

Reply:
happy to jump on a call there just ping me on slack

Reply:
Yes, the way it is implemented now is fine 👍  (empty keys, but if they are not present in the beginning there is no error and they will be added as empty keys before auto configuration is done).
Tom's concern was more about supporting two versions of what `config.yml` files can look like. In the current implementation, after initial training with auto config the file will always have the same format (empty key present with commented out configuration), so we're all good.